### ISSUE TITLE
Citus version upgrade policy: parent issue

### ISSUE THREAD

**Issue by ozgune** [citusdata/citus#1052]
We opened a few issues related to Citus upgrades (#1036, #980, #907) and also talked about this in internal emails.

I'm opening this issue to track the different points that we tracked. The following are related but still independent points (will update with input).

(1) ``apt-get update`` or ``yum update`` shouldn't break installed software. @jasonmp85 is scoping out this work and will open related issues in our packages directory.

(2) Have multiple versions of Citus installed at the same time. @fdr mentioned this requirement as part of #1036. He also mentioned in an internal email thread that there may be workarounds. Once he knows more, we'll add notes to this issue.

(3) What's our major and minor release update policy?

Until 5.0, we had the policy that each minor release would be backwards compatible (4.1 would be backwards compatible with 4.0). This practically meant two things:

* We don't make changes to system catalogs across minor releases
* We don't delete user-defined function definitions or change them across minor versions

I'm going over our citus--x.x.sql files in our [source code](https://github.com/citusdata/citus/tree/master/src/backend/distributed) and we seem to have mostly followed this policy.

I only see [this system catalog change](https://github.com/citusdata/citus/blob/master/src/backend/distributed/citus--5.2-3--5.2-4.sql) across minor version upgrades. We introduced this change as part of v6.0, so this looks like bad numbering.

For user-defined functions, we seem to have only changed them as part of schema support (#122) and mostly paid attention to making these changes minor (#691). As such, our policy on backwards compatibility doesn't seem to be hurting our development pace a lot. If I'm missing out on any other changes, could you update this thread?

(4) How far back do we maintain releases? Do we have a policy on this?

(5) What documentation do we provide to users for Citus / Postgres upgrades?

We previously offered the following instructions for upgrade Citus versions. These steps included both major upgrades from Citus 4.x to Citus 5 and also minor version upgrades across Citus versions: https://docs.citusdata.com/en/v5.2/admin_guide/upgrading_citus.html

In short, these steps ask pg_upgrade and metadata dump for a major version upgrade. They document a simple upgrade and ``ALTER EXTENSION citus UPDATE;``. If we have other expectations from users, we should capture them in our docs.

More importantly, we currently have private Citus 5.x to 6.x upgrade steps. We opened an issue in our docs repo to capture this and adding these steps would help make these steps more transparent: https://github.com/citusdata/citus_docs/issues/213

**Comment by ozgune** [comment]
We had an internal meeting on this and defined our policy on backwards compatibility. (question 3 above). I'm copy/pasting our policy for 6.2 and onwards from @sumedhpathak's email below.

--

I wanted to drop an email to document our versioning policy, and what we mean by major/minor/patch releases. Some of this has been understood but implicit, and other implicit policies we've followed in the past are now changing.

**Patch release:**
e.g. 6.0.0 -> 6.0.1. These releases *only* contain bug fixes which don't add new functionality, and don't modify catalog tables or function API's (or add new function API's).

Users will get these updates by:
- yum update
- restart backend

**Minor release:**
e.g. 6.1 -> 6.2.
This release contains new functionality, and *can* contain catalog table or API changes as long as they are additive and don't break previous functionality. A change which we consider as breaking this policy would be removal of a catalog table or GUC.

User's will need to explicitly update via:
- yum install citus-6.2
- restart backend
- 'alter extension citus update'.

**Major release:**
e.g. 6.3 -> 7.0
This release will contain new functionality, and also possibly breaking functionality which requires more effort from a user to upgrade. We may possible remove or change type of catalog table columns, remove catalog tables, change API signatures or add/remove GUC's in this change.

User's will need to explicitly update via:
- pre (and possibly post) prep/changes on the application side. 
- yum install citus-7.0
- restart backend server
- 'alter extension citus update'.

The one change in policy here is allowing catalog changes (as long as they are additive) in minor-versions (e.g. 6.1 -> 6.2). This is something we haven't done in the past, but are now formally allowing w.r.t our new policies.
